src/object.jl

```diff
@@ -47,43 +50,199 @@ function Base.unsafe_copyto!(dest::AbstractVector{UInt8}, doff::Integer, src::Ob
     return n
 end
 
-function Base.unsafe_copyto!(dest::Channel, doff::Integer, src::Object, soff::Integer, n::Integer)
-    bytes = getRange(src, soff, n)
-    dest_arr, new_n = take!(dest)
-    new_n == n || throw(ArgumentError("prefetched number of bytes (`$n`) doesn't matched requested number of bytes (`$new_n`)"))
-    copyto!(dest_arr::AbstractVector{UInt8}, doff, bytes)
-    put!(dest, n)
-    return n
+mutable struct TaskCondition
+    cond_wait::Threads.Condition
+    ntasks::Int
+end
+TaskCondition() = TaskCondition(Threads.Condition(), 0)
+mutable struct PrefetchBuffer
+    data::Vector{UInt8}
+    pos::Int
+    len::Int
+end
+Base.bytesavailable(b::PrefetchBuffer) = b.len - b.pos + 1
+
+function _prefetching_task(io)
+    prefetch_size = io.prefetch_size
+    len = io.len
+    ppos = 0
+    download_buffer = Vector{UInt8}(undef, prefetch_size)
+    # Don't allocate the second buffer when we can fit everything to the first one
+    download_buffer_next = len == prefetch_size ? download_buffer : Vector{UInt8}(undef, prefetch_size)
+
+    try
+        while len > ppos
+            n = min(len - ppos, prefetch_size)
+            off = 0
+            # `partition` plays nicely with 1-based ranges, but we need zero based ranges for
+            # `contentRange`
+            rngs = Iterators.partition(ppos+1:ppos+n, io.prefetch_multipart_size)
+            io.cond.ntasks = length(rngs)
+            for rng in rngs
+                put!(io.download_queue, (off, rng .- 1, download_buffer))
+                off += length(rng)
+            end
+            Base.@lock io.cond.cond_wait begin
+                while true
+                    io.cond.ntasks == 0 && break
+                    wait(io.cond.cond_wait)
+                end
+            end
+
+            buf = PrefetchBuffer(download_buffer, 1, n)
+            ppos += n
+            put!(io.prefetch_queue, buf)
+            download_buffer, download_buffer_next = download_buffer_next, download_buffer
+        end
+    catch e
+        close(io.download_queue, e)
+        close(io.prefetch_queue, e)
+        rethrow()
+    end
+    close(io.download_queue)
+    close(io.prefetch_queue)
+    return nothing
+end
+
+function _download_task(io)
+    headers = HTTP.Headers()
+    object = io.object
+    url = makeURL(object.store, io.object.key)
+    credentials = object.credentials
+    response_stream = IOBuffer(view(UInt8[], 1:0), write=true, maxsize=io.prefetch_multipart_size)
+
+    try
+        while true
+            (off, rng, download_buffer) = take!(io.download_queue)
+            HTTP.setheader(headers, contentRange(rng))
+            buffer_view = view(download_buffer, off + 1:off + length(rng))
+            response_stream.data = buffer_view
+            response_stream.maxsize = length(buffer_view)
+            seekstart(response_stream)
+            _ = getObject(object.store, url, headers; credentials, response_stream)
```

Comment:
Note for this to work properly without allocating, we need to fix IOBuffer handling in HTTP.jl.

Reply:
https://github.com/JuliaWeb/HTTP.jl/pull/987